File: java/src/main/java/benchmark/serializers/ProtostuffSer.java
package benchmark.serializers;

import benchmark.model.Fixture;
import io.protostuff.LinkedBuffer;
import io.protostuff.ProtostuffIOUtil;
import io.protostuff.Schema;
import io.protostuff.runtime.RuntimeSchema;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Protostuff runtime — protobuf-style binary without .proto for POJOs (popular RPC path).
 *
 * <p>Recommended: cache {@link RuntimeSchema} per type; reuse {@link LinkedBuffer}; use
 * {@link ProtostuffIOUtil#toByteArray}/{@link ProtostuffIOUtil#mergeFrom} for singles and
 * {@code writeListTo}/{@code parseListFrom} for batches.
 *
 * @see <a href="https://github.com/protostuff/protostuff">protostuff</a>
 */
public final class ProtostuffSer implements BenchSerializer {
  private final LinkedBuffer buffer = LinkedBuffer.allocate(LinkedBuffer.DEFAULT_BUFFER_SIZE);
  private Schema<Object> schema;
  private boolean batch;
  private Class<?> elementClass;

  @Override
  public String name() {
    return "protostuff";
  }

  @Override
  public String version() {
    return Versions.of(ProtostuffIOUtil.class);
  }

  @Override
  public String streamMode() {
    return "native";
  }

  @Override
  public String nativeKind() {
    return "schema";
  }

  @Override
  @SuppressWarnings("unchecked")
  public void prepare(Fixture fx) {
    batch = TypeUtil.isList(fx.value);
    elementClass = TypeUtil.elementClass(fx.value);
    schema = (Schema<Object>) (Schema<?>) RuntimeSchema.getSchema(elementClass);
    buffer.clear();
  }

  @Override
  @SuppressWarnings("unchecked")
  public byte[] serializeBytes(Fixture fx) throws Exception {
    buffer.clear();
    if (batch) {
      ByteArrayOutputStream baos = new ByteArrayOutputStream(4096);
      ProtostuffIOUtil.writeListTo(baos, (List<Object>) fx.value, schema, buffer);
      return baos.toByteArray();
    }
    return ProtostuffIOUtil.toByteArray(fx.value, schema, buffer);
  }

  @Override
  public Object deserializeBytes(byte[] data) throws Exception {
    if (batch) {
      return ProtostuffIOUtil.parseListFrom(new ByteArrayInputStream(data), schema);
    }
    Object msg = schema.newMessage();
    ProtostuffIOUtil.mergeFrom(data, msg, schema);
    return msg;
  }

  @Override
  @SuppressWarnings("unchecked")
  public int serializeStream(Fixture fx, OutputStream out) throws Exception {
    buffer.clear();
    if (batch) {
      return ProtostuffIOUtil.writeListTo(out, (List<Object>) fx.value, schema, buffer);
    }
    return ProtostuffIOUtil.writeTo(out, fx.value, schema, buffer);
  }

  @Override
  public Object deserializeStream(InputStream in) throws Exception {
    if (batch) {
      return ProtostuffIOUtil.parseListFrom(in, schema);
    }
    Object msg = schema.newMessage();
    ProtostuffIOUtil.mergeFrom(in, msg, schema);
    return msg;
  }
}

File: kotlin/src/main/kotlin/benchmark/serializers/ProtostuffSer.kt
package benchmark.serializers

import benchmark.model.Fixture
import io.protostuff.LinkedBuffer
import io.protostuff.ProtostuffIOUtil
import io.protostuff.Schema
import io.protostuff.runtime.RuntimeSchema
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.OutputStream

/**
 * Protostuff runtime — protobuf-style POJO binary without .proto.
 *
 * Hot path: cache [RuntimeSchema]; reuse [LinkedBuffer]; list APIs for batches.
 */
class ProtostuffSer : BenchSerializer {
    private val buffer = LinkedBuffer.allocate(LinkedBuffer.DEFAULT_BUFFER_SIZE)
    private val baos = ByteArrayOutputStream(4096)
    private lateinit var schema: Schema<Any>
    private var batch = false

    override fun name() = "protostuff"

    override fun version() = Versions.of(ProtostuffIOUtil::class.java)

    override fun streamMode() = "native"

    override fun nativeKind() = "schema"

    @Suppress("UNCHECKED_CAST")
    override fun prepare(fx: Fixture) {
        batch = TypeUtil.isList(fx.value)
        schema = RuntimeSchema.getSchema(TypeUtil.elementClass(fx.value)) as Schema<Any>
        buffer.clear()
    }

    @Suppress("UNCHECKED_CAST")
    override fun serializeBytes(fx: Fixture): ByteArray {
        buffer.clear()
        if (batch) {
            baos.reset()
            ProtostuffIOUtil.writeListTo(baos, fx.value as List<Any>, schema, buffer)
            return baos.toByteArray()
        }
        return ProtostuffIOUtil.toByteArray(fx.value, schema, buffer)
    }

    override fun deserializeBytes(data: ByteArray): Any {
        if (batch) {
            return ProtostuffIOUtil.parseListFrom(ByteArrayInputStream(data), schema)
        }
        val msg = schema.newMessage()
        ProtostuffIOUtil.mergeFrom(data, msg, schema)
        return msg
    }

    @Suppress("UNCHECKED_CAST")
    override fun serializeStream(fx: Fixture, out: OutputStream): Int {
        buffer.clear()
        return if (batch) {
            ProtostuffIOUtil.writeListTo(out, fx.value as List<Any>, schema, buffer)
        } else {
            ProtostuffIOUtil.writeTo(out, fx.value, schema, buffer)
        }
    }

    override fun deserializeStream(input: InputStream): Any {
        if (batch) {
            return ProtostuffIOUtil.parseListFrom(input, schema)
        }
        val msg = schema.newMessage()
        ProtostuffIOUtil.mergeFrom(input, msg, schema)
        return msg
    }
}
